src/Stratis.Bitcoin/Connection/ConnectionManager.cs

```diff
@@ -257,6 +259,41 @@ private void AddComponentStats(StringBuilder builder)
                     + " agent:" + agent);
             }
 
+            var oneTryPeers = this.ConnectedPeers.Where(p => p.Behavior<ConnectionManagerBehavior>().OneTry);
```

Comment:
Is it a good idea to iterate ConnectedPeers 3 times?

Reply:
In this case I would say its fairly innocuous. Typically there are not that many items so setting up two additional string builders may be overkill. If you want I can easily change it.

Reply:
`.Behavior<ConnectionManagerBehavior>()` also does a FirstOrDefault() to get the correct behaviour. Considering we could have ~100 peers which could have several Behaviours, and will be doing this every 5 seconds, maybe there is a way we can store them in some kind of mapping so we only do this work once?

Reply:
I have made some changes to only iterate ConnectedPeers once. Created a dictionary for AddNode checking. Further optimization is certainly possible if we regard this important enough.